### PR TITLE
Adding the option to convert bidirectionally between map type (Protobuf) to list of structs (Arrow).

### PR DESCRIPTION
**Pull request by guillaume-rochette-oxb**
Hi,

I have a use case, where I convert bidirectionally collections of Protobuf messages to Arrow data structures with Protarrow, before writing to [Parquet](https://arrow.apache.org/docs/python/parquet.html) datasets and inserting into tables in [BigQuery](https://docs.cloud.google.com/bigquery/docs/reference).
However BigQuery does not have a native [map type](https://docs.cloud.google.com/bigquery/docs/reference/standard-sql/data-types).
Instead, it implicitly converts a `m: Map<k: K, v: V>` to a `m: Struct<l: List<e: Struct<k: K, v: V>>> ` at insertion time, however it does not retain the original schema, and therefore does not revert to the original the data structure when exporting back to Parquet datasets, which breaks the return trip from Arrow back to Protobuf.

At first, I was thinking to make a separate Python package that would compare the original schema derived from the Protobuf messages and the schema embedded in the RecordBatch in a depth-first-search manner, in order to convert back to what Protarrow expects.
But I figured that it might be useful to the Protarrow community, and also that performing two depth-first-search back-to-back conversions would be wasteful when applied at a scale, especially in my use case where the Parquet/Arrow representations are only intermediate and temporary representations.

Please note that instead of mimicking the implicit conversion operated by BigQuery from `m: Map<k: K, v: V>` to `m: Struct<l: List<e: Struct<k: K, v: V>>> `, it maps to `l: List<e: Struct<k: K, v: V>> ` for two reasons, 1) it follows the underlying map representation from the Protobuf [specifications](https://protobuf.dev/programming-guides/encoding/#maps), and 2) because BigQuery limits the depth of a table schema to 15, c.f. their [documentation](https://docs.cloud.google.com/bigquery/docs/nested-repeated#limitations), so might save some people a headache in the future.

I have tried to the make the changes as minimally invasive as possible, and the computational complexity is not affected.
I have added a new config to the list in the tests and everything was passing.

Thanks in advance,
Guillaume

### CODE REVIEW

**Comment by 0x26res** [comment]
Hi @guillaume-rochette-oxb 

First, thanks for your contribution.

Before I look more closely at the code, I'd like to understand the use case. So let me rephrase what you said to make sure I understand it. I apologize in advance for being inquisitive.

One can inject parquet files directly into BigQuery. BigQuery doesn't support maps. Therefore they get converted to a `pa.struct(pa.field('???', pa.list(pa.struct(pa.field('???', <KeyType>), pa.field('???', <ValueType>)`. For some reason there's an extra/unnecessary level of nesting here. 
- does this conversion from map to nested list of struct  happen automatically? 
- Also could you confirm the name of the fields?

And then I guess the other way around would be to either export a query to parquet gcs or to just query the table directly, and the schema would mismatch since there is no map. But a struct containing a list of struct.

- Is your workflow Protobuf -> Arrow -> Parquet -> BigQuery followed by BiqQuery -> Parquet -> Arrow -> Protobuf?

I have a little experience with the BigQuery export to parquet format and it has given me problems before around timestamp precision and timezone (I think precision is truncated and the UTC timezone are dropped from TIMESTAMP). I guess it's not a problem for you given you provide the protobuf schema when you read back the data and it get interpreted correctly.

But all this is to say that the BigQuery to proto export is a bit of a can of worms because of such edge case. And supporting it would make protarrow more convoluted than it should be

At a first glance, I would argue for not trying to support high fidelity conversion to bigquery, because their convention is opinionated. Instead we could add a flag, say `map_as_list`, which would save map as a list of struct (and also do the right convention when casting). And then keep the code you've added in arrow_to_proto (this one auto detects list of structs and doesn't need a flag). We may also support the hedge case where the list is nested. We should document somewhere the fact that this "magic" happens.

A couple side notes:
1. there is an API to upload protos directly to bigquery. I guess it's probably less relevant if you are dealing with big batches. https://docs.cloud.google.com/dotnet/docs/reference/Google.Cloud.BigQuery.Storage.V1/latest/Google.Cloud.BigQuery.Storage.V1.BigQueryWriteClient.AppendRowsStream. I'm not sure if you can read the data back with this, as protobuf.
2. You mentioned a worry about doing 2 depth-search in the schema. I would not worry about it, this is not expensive and could possibly be cached. Also even the operation of converting the list of struct to a map is cheap because the underlying data isn't copied. It is just changing the meta data (if done correctly). BTW If you care about performance, check out ptars https://github.com/0x26res/ptars

**Comment by guillaume-rochette-oxb** [comment]
Hi,

Yes, the conversion is done implicitly by the [LOAD DATA](https://docs.cloud.google.com/bigquery/docs/reference/standard-sql/load-statements) statement.
But it does not happen in the [EXPORT DATA](https://docs.cloud.google.com/bigquery/docs/reference/standard-sql/export-statements) statement, because the original schema is not retained.

For the naming, if the input `pa.Schema` is:
```
m: map<int64, string>
  child 0, entries: struct<key: int64 not null, value: string> not null
      child 0, key: int64 not null
      child 1, value: string
```
The output `pa.Schema`, e.g. after write to Parquet, load to BigQuery, export from BigQuery and read from Parquet is:
```
m: struct<key_value: list<element: struct<key: int64, value: string> not null> not null>
  child 0, key_value: list<element: struct<key: int64, value: string> not null> not null
      child 0, element: struct<key: int64, value: string> not null
          child 0, key: int64
          child 1, value: string
```
We can see that there's one extra level of nesting, (which is also there in the table schema), and they could have done the following and spared one level of nesting:
```
m: list<element: struct<key: int64, value: string> not null> not null
  child 0, element: struct<key: int64, value: string> not null
      child 0, key: int64
      child 1, value: string
```

Yes, that is the use case, guaranteeing the roundtrip of Protobuf messages is crucial.

Yes, the [TIMESTAMP](https://docs.cloud.google.com/bigquery/docs/reference/standard-sql/data-types#timestamp_type) in BigQuery have a microsecond precision and are necessarily in UTC, but that is not an issue in my case.

With the current PR, we achieve 100% message roundtrip equality between Protobuf and BigQuery for the `ExampleMessage`, `NestedExampleMessage`, `SuperNestedExampleMessage`. I had to touch up the `tests/random_generator.py` for generating timestamps with a microsecond precision, and I used the following `config`:
```
p2a.ProtarrowConfig(
    timestamp_type=pa.timestamp(unit="us", tz="UTC"),
    time_of_day_type=pa.time64(unit="us"),
    duration_type=pa.duration(unit="us"),
    list_nullable=True,
    map_nullable=True,
    skip_recursive_messages=True,
    map_as_list_of_structs=True,
)
```

Yes, I am happy to rename from `map_as_list_of_structs` to `map_as_list`, shorter is fine.

Sorry, I did not understand what you meant,

> [...], which would save map as a list of struct (and also do the right convention when casting). And then keep the code you've added in arrow_to_proto (this one auto detects list of structs and doesn't need a flag). We may also support the hedge case where the list is nested. We should document somewhere the fact that this "magic" happens.

As for the side notes,
1. I want to avoid streaming into tables in general, it complexifies things, and forces me to deal with problems like late arrival, etc.
2. I will have a look at `ptars`, that much speed-up sounds appealing :)

File: protarrow/arrow_to_proto.py [hunk 1]
```diff
@@ -458,10 +458,25 @@ def _extract_struct_field(
...
-    array: pa.MapArray,
+    array: Union[pa.MapArray, pa.ListArray, pa.LargeListArray],
     field_descriptor: FieldDescriptor,
     messages: Iterable[Message],
 ) -> None:
+
+    if pa.types.is_list(array.type) or pa.types.is_large_list(array.type):
+        assert pa.types.is_struct(array.values.type), array.values.type
+        field_names = [field.name for field in array.values.type.fields]
+        assert len(field_names) == 2
```
Comment (0x26res) [inline protarrow/arrow_to_proto.py]: Can you add a more explicit error messages, raising exceptions. Do it before extracting field names, basically we are checking the dtype is valid before doing any transformation.

Also I think we should assume the key is first and value next. So rely on the order of field rather than having a field called `key` the other being `value` by elimination.

Lastly I would put this outside in a private method, and add a doc string explaining it's for the map as list use case.

File: protarrow/cast_to_proto.py [hunk 2]
```diff
@@ -107,26 +107,73 @@ def _cast_array(
...
+        if pa.types.is_map(array.type):
+            keys = array.keys
+            values = array.items
+        else:
+            assert pa.types.is_list(array.type) or pa.types.is_large_list(array.type), (
+                array.type
+            )
+            assert pa.types.is_struct(array.values.type), array.values.type
+            keys = array.values.field("key")
+            values = array.values.field(config.map_value_name)
```
Comment (0x26res) [inline protarrow/cast_to_proto.py]: The use of map_value_name isn't consistent on the code above. I would argue again we should just rely on the field index (key=0, value=1).

Also can you make the checks throw proper errors again. Possibly add a helper for the check or a helper to extract the key and value arrays.

File: protarrow/cast_to_proto.py [hunk 3]
```diff
@@ -107,26 +107,73 @@ def _cast_array(
...
+
+        # TODO: remove when https://github.com/apache/arrow/issues/40750 is fixed
+        #  and library is pinned to pyarrow>=17.0.0
```
Comment (0x26res) [inline protarrow/cast_to_proto.py:128]: That's a good point, I've created an issue https://github.com/tradewelltech/protarrow/issues/125 

I'll remove this once you merge your PR but lest leave this here for now.

File: protarrow/cast_to_proto.py [hunk 4]
```diff
@@ -107,26 +107,73 @@ def _cast_array(
...
+        offsets = maybe_copy_offsets(array.offsets)
+        keys = _cast_array(keys, key_field, config)
+        values = _cast_array(values, value_field, config)
+
+        if config.map_as_list:
+            return config.list_array_type.from_arrays(
```
Comment (0x26res) [inline protarrow/cast_to_proto.py]: This is a mouthful can we put it in a private function?

File: protarrow/cast_to_proto.py [hunk 5]
```diff
@@ -107,26 +107,73 @@ def _cast_array(
...
+                            nullable=False,
+                        ),
+                        pa.field(
+                            name=config.map_value_name,
+                            type=values.type,
+                            nullable=config.map_value_nullable,
+                        ),
+                    ],
                 ),
+            ).cast(
```
Comment (0x26res) [inline protarrow/cast_to_proto.py]: why do we need the cast? If it's really needed can you add a comment?

File: protarrow/proto_to_arrow.py [hunk 6]
```diff
@@ -498,14 +517,54 @@ def _proto_map_to_array(
...
+                            nullable=False,
+                        ),
+                        pa.field(
+                            name=config.map_value_name,
+                            type=values.type,
+                            nullable=config.map_value_nullable,
+                        ),
+                    ]
+                )
+            )
```
Comment (0x26res) [inline protarrow/proto_to_arrow.py]: same comment as above regarding the cast / and can we share the code?

File: tests/test_coverage.py [hunk 7]
```diff
@@ -61,6 +62,19 @@ def test_map_converter_adapter():
...
+    map_converter_adapter = MapAsListConverterAdapter(
+        list_type=list_type,
+        key_descriptor=map_field.message_type.fields_by_name["key"],
+        value_descriptor=map_field.message_type.fields_by_name["value"],
+    )
+    assert map_converter_adapter(pa.scalar([(123, 1.0)], list_type)) == {123: 1.0}
+    assert map_converter_adapter(pa.scalar([], list_type)) == {}
+    assert map_converter_adapter(pa.scalar(None, list_type)) == {}
+
+
```
Comment (0x26res) [inline tests/test_coverage.py:81]: We need to add more tests here. I know the test suite makes it look like we achieve 100% coverage. But we want to check that the arrow record batch's schema is what we expect when we change the behavior.

Also there's a few corner cases to test.

Here are the tests:
- Create 3 `ExampleMessage`. Populate one of its map field (eg double_int32_map) with a couple of element, one element and no element (respectively). Convert the messages to arrow record batch with the map_as_list flag on. Check that the resulting schema is what you expected, and that the values in the array (when converted to as_py) are correct. Possibly parameterize this test on a custom map_value_name.
- Create a record batch with one column, that contains list of struct (double_int32_map). Convert this back to ExampleMessage (you don't need t worry about other fields). Check that it pick up the map content correctly. Also run this test with a large_list of struct.

Thanks.

**Review by 0x26res — COMMENTED** [review]
I've put a few comments it's mainly minor things.

File: protarrow/cast_to_proto.py (hunk 5 above)
Comment (guillaume-rochette-oxb) [reply protarrow/cast_to_proto.py]: We need the cast because we are using the `ProtarrowConfig.list_` method. Otherwise it does not roundtrip.

File: tests/test_coverage.py (hunk 7 above)
Comment (guillaume-rochette-oxb) [reply tests/test_coverage.py:81]: I do not understand why this is needed since, it roundtrips with the `map_as_list=True`, but I can add those.

**Comment by guillaume-rochette-oxb** [comment]
I think I resolved everything :crossed_fingers:

File: tests/test_coverage.py [hunk 8]
```diff
@@ -61,6 +66,88 @@ def test_map_converter_adapter():
...
+            5: 0.2,
+            2: 0.3,
+        },
+        {
+            6: 0.2,
+        },
+        {},
+    ]
+    message_batch = []
+    for data in data_batch_0:
```
Comment (0x26res) [inline tests/test_coverage.py]: please don't use numbers for the variable, but names relevant to their context. 

data_batch_0 -> input_data, data_batch_1 -> output_data

File: tests/test_coverage.py (hunk 7 above)
Comment (0x26res) [reply tests/test_coverage.py:81]: > I do not understand why this is needed since, it roundtrips with the map_as_list=True, but I can add those.

as mentioned:

> But we want to check that the arrow record batch's schema is what we expect when we change the behavior.

the roundtrip test don't check the intermediate schema.

**Comment by guillaume-rochette-oxb** [comment]
@0x26res Is there anything else that needs to be checked?

**Review by 0x26res — APPROVED** [review]

**Comment by guillaume-rochette-oxb** [comment]
Thanks :) When do you think the next release will happen?

**Comment by 0x26res** [comment]
It is released now https://github.com/tradewelltech/protarrow/releases/tag/v0.16.0
